### ISSUE TITLE
Troubleshooting poor prediction and vectorisation output

### ISSUE THREAD

**Issue by SFrav** [sentinel-hub/field_delineation#14]
We have applied this workflow using over 500k reference polygons from a region of Spain. The input data is high-quality. The predictions and resulting vector data is poor.

What can we do to improve performance?

We have 120 EO-patches of 1200x1200 px each. Training was done with 400 iterations and 25 epochs (1500 iterations had minor improvements in losses after 1 fold). We tried a range of threshold values for vectorisation.
We have also tried making improvements to the reference data, including removing small polygons, removing slithers (area to length ratio) and using more specific land use categories (e.g. 300k polygons of annuals). 
Restricting reference data to annual crop polygons does improve training losses.

The issues are as follows:
- there are clear EO-patch specific differences in predicted pseudo probabilities and so vecotrisation
- over-segmentation in areas outside of the reference data extent (i.e. where there is no agriculture
- identifies shapes not in the reference data, such as rivers and buildings
- in a few areas there are some polygons that somewhat resemble the reference data. None of these are high quality matches

A few questions:
- what more can be done to improve EO-patch specific differences
- what is the minimum volume of reference data needed? Is 500k close?
- is any further pre-processing of reference data needed? Beyond size and shape?
- what are the loss values for your best performing model? See our best performance [here](https://github.com/sentinel-hub/field-delineation/files/8671868/training_metrics_annuals_1500.csv)
- how would we go about producing predictions for extent, boundary and distance pseudo probabilities? This might be useful in identifying alternative vectorisation/post-processing approaches
- have you tested different EO input data, such as change in band values over time? e.g B02_t2 - B02_t1 etc.

**Comment by devisperessutti** [comment]
Hi @SFrav ,

Sorry to hear results are not (yet) good. Could you provide some plots of bands/predicted extent/predicted boundary and possibly resulting vector, to get a feeling for what is going on?

>  there are clear EO-patch specific differences in predicted pseudo probabilities and so vecotrisation

We didn't notice this, the vectors over neighbouring eopatches match well. I think the problem is the model hasn't converged yet.
   
> over-segmentation in areas outside of the reference data extent (i.e. where there is no agriculture

This can happen, especially if you provide only agricultural fields to the model at training, since it didn't see other land cover examples, therefore not knowing how to correctly classify them. To alleviate this, add negative samples to your training dataset that include non-agricultural land cover.

> identifies shapes not in the reference data, such as rivers and buildings

Same issue as above, add negative samples (rivers/buildings) to training data.

> in a few areas there are some polygons that somewhat resemble the reference data. None of these are high quality matches

Could you provide such cases?

Looking at the iteration scores, your training losses converge while your validation losses don't. I would therefore double-check that train/val datasets are correct, and make predictions on both training and validation dataset. If there are no issues, you might want to lower your learning rate, or possibly reduce the number of features in the layers to simplify the complexity.

It sounds like your training data is large enough to get some initial results, so I don't think it's an issue.

> what are the loss values for your best performing model?

We report these in the blog post, they are similar to your training losses.

> is any further pre-processing of reference data needed? Beyond size and shape?

We didn't remove any reference vector up-front.

> how would we go about producing predictions for extent, boundary and distance pseudo probabilities?

Unless you modified the code, the prediction step generates all three probability arrays (extent, boundary, distance) and saves them to the EOPatch. From here you could combine them as you wished. I suggest visualising some of these probas.

We fine-tuned our original model for some areas in Castilla and Andalusia and results were similar to the ones shown in [parcelio](http://parcelio.sentinel-hub.com/).

You can send us an EOPatch with predictions and some sample npz files, but I can't promise when/if we'll manage to look at them. Best if you attach here some figure examples.

**Comment by SFrav** [comment]
Hi @devisperessutti, thanks for the quick reply!

An example EOPatch, along with raw EO data, vector output, npzs and an average model is available [here](https://drive.google.com/drive/folders/1ZmCVnm64DmWn4E5U2qd7Unozx8Q6e86-?usp=sharing).

Here's an example of the predicted pseudo probs:
![](https://user-images.githubusercontent.com/6743843/168119294-1ec90aa0-3564-4760-8775-865a513dca89.jpg)

Here's an example of one of the better quality vector outputs. In this example there is a river predicted as a field.
![](https://user-images.githubusercontent.com/6743843/168119690-70a7cea0-3303-463d-bceb-69b05211c4a5.JPG)


A few follow-up questions (to show my ignorance):
- Is [this old notebook](https://github.com/sentinel-hub/field-delineation/blob/3e9a493bdfa7146b66abd36a844618250a77e190/notebooks/04-patchlets-sampling.ipynb) a good reference for adding the negative samples. Perhaps with fewer num_samples?
- How would I go about checking train/val datasets? 
- How would I predict on the validation dataset? I don't see anything in prediction.py or the script

**Comment by devisperessutti** [comment]
Yes, results don't make much sense. The predicted extent and boundary maps should look like the ones provided as reference.

Looking at these results, I would double-check that `softmax` is applied to the estimated arrays, and the correct logits are used in the calculation of the loss. You can check this by looking at the two channels of the predicted arrays (extent/boundary) and verify they sum to 1.

Which `eo-flow` version/branch are you using? I would suggest using the `develop` branch.

Your patchlets look OK otherwise. I would suggest to randomly shuffle them even more before creating npz files, as there were few samples from same eopatch in sequence (this should not be the main cause of the problem though). 


> Is [this old notebook](https://github.com/sentinel-hub/field-delineation/blob/3e9a493bdfa7146b66abd36a844618250a77e190/notebooks/04-patchlets-sampling.ipynb) a good reference for adding the negative samples. Perhaps with fewer num_samples?

Yes, correct. Negative samples should be a minority of the total, as we want the model to mainly learn to delineate parcels.   

> How would I go about checking train/val datasets?

Once you have tensorflow datasets for each train/val fold, plot some samples out to check they make sense, in particular when applying augmentations.

>  How would I predict on the validation dataset? I don't see anything in prediction.py or the script

You might have to check some older training notebook, where it shows how you can call `model.net.predict(validation_dataset)`

I think you are close to get it to work, try to debug different steps making sure is what is expected, and try tweaking the learning rate.

Good luck :muscle:

**Comment by SFrav** [comment]
Great!

The eo-flow version is 1.2.0. I'll try with the dev version and check the recent commit to this repo.

For the negative samples, this would create a new folder, called say patchlets-neg. I would then run patchlets_to_npz_files on both folders, right? This wasn't clear in that older workbook.

Now time to check the predicted arrays, train/val datasets and predict on the validation dataset.

Thanks 💪

**Comment by devisperessutti** [comment]
> For the negative samples, this would create a new folder, called say patchlets-neg. I would then run patchlets_to_npz_files on both folders, right? This wasn't clear in that older workbook.

Yes, that is a good option. If you wanted to have more control over how many negative samples you add, you could create a separate tensorflow dataset and add a given fraction to your training dataset.

What I would do is try to get the model to work first and then add negative samples, since to me it seems the main issues lie in the model.

**Comment by SFrav** [comment]
> double-check that softmax is applied to the estimated arrays, and the correct logits are used in the calculation of the loss. You can check this by looking at the two channels of the predicted arrays (extent/boundary) and verify they sum to 1.

The sums range from 0.98 to 1.01. Seems OK.

>Once you have tensorflow datasets for each train/val fold, plot some samples out to check they make sense

I'll have to look into this. Not in tensorboard so will check if the callback in `ExtentBoundDistVisualizationCallback` (tf_viz_utils.py) is being used.

>you can call model.net.predict(validation_dataset)

Still to do.

>try tweaking the learning rate

I've been using 0.0001. I'll try 0.00005 on the next model run and will keep an eye out for overfitting.

**Comment by SFrav** [comment]
OK

Learning rates of 0.0005 and 0.005 didn't improve convergence. That's from looking at the validation losses. If anything, 0.005 leveled off sooner, which may suggest overfitting. See losses as follows: 
[Train_00005.csv](https://github.com/sentinel-hub/field-delineation/files/8702295/Train_00005.csv)
[Train_005.csv](https://github.com/sentinel-hub/field-delineation/files/8702296/Train_005.csv)

Predicting on validation set differed for a single model vs the average model. See below. Code for plotting is [here](https://github.com/SFrav/field-delineation/blob/7f00bfa383346bd7800b4b547a7759739456e288/plot_predict_validation.py) in case others can make use of it.

Single model:
![predict_val_1](https://user-images.githubusercontent.com/6743843/168654318-5e98184a-f8fb-4c28-8734-744b75aee0e0.jpg)
 
Average model:
![predict_val_ave](https://user-images.githubusercontent.com/6743843/168654373-23882b8c-5dea-453a-ba84-0b1f67fca681.jpg)

Next I'll try using the dev version of eo-flow.

Is there anything I'm missing in interpreting the losses and prediction on validation set?

**Comment by devisperessutti** [comment]
Hi @SFrav 

Few Qs:

 * what is the total number of samples you end up with?
 * what is the number of iterations per epoch, and number of epochs?

It still looks that while the training loss is slowly converging, the validation loss is not. So there might be too few iterations per epoch, or epochs.

**Comment by SFrav** [comment]
@devisperessutti 

I'm not sure what you mean by total number of samples. For each fold there are between 190 and 240 npz files.

The models above were run with 400 iterations and 25 epochs. I'll try more iterations on this next run.

**Comment by SFrav** [comment]
Increasing the iterations to 1500 and using the dev branch of eo-flow seems to be working.
[Losses here](https://github.com/sentinel-hub/field-delineation/files/8732508/Training_metrics_1500.csv).

This particular run cut out before the end of the first fold. I'll try again and close this issue for now.
.
